box/src/main/scala/blobstore/box/BoxStore.scala

```diff
@@ -0,0 +1,238 @@
+          throw new Exception(s"Item '${matchingItem.getName}' exists along path but was not folder")
+        } else {
+          val folder = matchingItem.getResource.asInstanceOf[BoxFolder]
+          putFolderAtPath(folder, tail)
+        }
+    }
+  }
+
+  def splitPath(path: Path): (List[String], String) = {
+    val fullPath = path.root :: path.key.split("/").toList
+    val pathToParentFolder = fullPath.dropRight(1)
+    val key = fullPath.takeRight(1).head
+    (pathToParentFolder, key)
+  }
+
+  /**
+    * Provides a Sink that writes bytes into the provided path. See [[StoreOps.PutOps]] for convenient put String
+    * and put file methods.
+    *
+    * It is highly recommended to provide [[Path.size]] when writing as it allows for optimizations in some store.
+    *
+    * @param path to put
+    * @return sink of bytes. This throws an exception if a file at this path already exists.
+    */
+  override def put(path: Path): Sink[F, Byte] = { in =>
+    val pathSplit = splitPath(path)
+
+    val init: F[(OutputStream, InputStream, BoxFolder)] = F.delay {
+      val parentFolder = putFolderAtPath(rootFolder, pathSplit._1)
+      val os = new PipedOutputStream()
+      val is = new PipedInputStream(os)
+      (os, is, parentFolder)
+    }
+
+    val consume: ((OutputStream, InputStream, BoxFolder)) => Stream[F, Unit] = ios => {
+      val putToBox = Stream.eval(F.delay({
+        ios._3.uploadFile(ios._2, pathSplit._2)
+        ()
+      }))
+      val writeBytes = _writeAllToOutputStream1(in, ios._1).stream ++ Stream.eval(F.delay(ios._1.close()))
+
+      putToBox concurrently writeBytes
+    }
+
+    val release: ((OutputStream, InputStream, BoxFolder)) => F[Unit] = ios => F.delay {
+      ios._2.close()
+      ios._1.close()
+    }
+
+    Stream.bracket(init)(consume, release)
+  }
+
+  /**
+    * Moves bytes from srcPath to dstPath. Stores should optimize to use native move functions to avoid data transfer.
+    *
+    * @param src path
+    * @param dst path
+    * @return F[Unit]
+    */
+  override def move(src: Path, dst: Path): F[Unit] = F.delay {
+    boxFileAtPath(src)
+      .foreach(file => {
+        val dstPath = splitPath(dst)
+        val folder = putFolderAtPath(rootFolder, dstPath._1)
+        file.move(folder, dstPath._2)
+      })
+  }
+
+  /**
+    * Copies bytes from srcPath to dstPath. Stores should optimize to use native copy functions to avoid data transfer.
+    *
+    * @param src path
+    * @param dst path
+    * @return F[Unit]
+    */
+  override def copy(src: Path, dst: Path): F[Unit] = F.delay {
+    boxFileAtPath(src)
+      .foreach(file => {
+        val dstPath = splitPath(dst)
+        val folder = putFolderAtPath(rootFolder, dstPath._1)
+        file.copy(folder, dstPath._2)
+      })
+  }
+
+  /**
+    * Remove byte for given path.
+    *
+    * @param path to remove
+    * @return F[Unit]
+    */
+  override def remove(path: Path): F[Unit] = F.delay {
+    boxFileAtPath(path).foreach(_.delete())
```

Comment:
should we be using handleWithError() here?

Reply:
we are letting error bubble up to users at this point, we may want to revise error handling across all stores.